clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp

```diff
@@ -304,121 +305,61 @@ static Expected<StringRef> linkDeviceLibFiles(StringRef InputFile,
-  CmdArgs.push_back(File);
-  if (Error Err = executeCommands(*LLVMToSPIRVProg, CmdArgs))
-    return std::move(Err);
-
-  if (!SPIRVDumpDir.empty()) {
-    std::error_code EC =
-        llvm::sys::fs::create_directory(SPIRVDumpDir, /*IgnoreExisting*/ true);
-    if (EC)
-      return createStringError(
-          EC,
-          formatv("failed to create dump directory. path: {0}, error_code: {1}",
-                  SPIRVDumpDir, EC.value()));
-
-    StringRef Path = OutputFile;
-    StringRef Filename = llvm::sys::path::filename(Path);
-    SmallString<128> CopyPath = SPIRVDumpDir;
-    CopyPath.append(Filename);
-    EC = llvm::sys::fs::copy_file(Path, CopyPath);
-    if (EC)
-      return createStringError(
-          EC,
-          formatv(
-              "failed to copy file. original: {0}, copy: {1}, error_code: {2}",
-              Path, CopyPath, EC.value()));
+/// be parsed to generate options required to be passed into the backend.
+static Expected<StringRef> runLLVMSPIRVBackend(StringRef File,
+                                               const ArgList &Args) {
+  llvm::TimeTraceScope TimeScope("LLVMToSPIRVTranslationViaSPIRVBackend");
+  std::vector<std::string> ExtNames, Opts;
+  if (Verbose || DryRun) {
+    errs() << formatv("LLVM-SPIRV-Backend: input: {0}, output: {1}\n", File,
+                      OutputFile);
+    if (DryRun)
+      return OutputFile;
   }
 
+  SMDiagnostic Err;
+  LLVMContext C;
+  std::unique_ptr<Module> M = parseIRFile(File, Err, C);
+  if (!M)
+    return createStringError(inconvertibleErrorCode(), Err.getMessage());
+
+  static const std::string DefaultTriple = "spirv64v1.6-unknown-unknown";
+
+  // Correct the Triple value if needed
+  // TODO: Remove this correction once we start using spirv64/spirv32 triples
+  // everywhere.
+  Triple TargetTriple(M->getTargetTriple());
+  if (TargetTriple.isSPIR()) {
+    TargetTriple.setArch(TargetTriple.getArch() == Triple::spir64
+                             ? Triple::spirv64
+                             : Triple::spirv32,
+                         TargetTriple.getSubArch());
+    M->setTargetTriple(TargetTriple);
+    // We need to reset Data Layout to conform with the TargetMachine
+    M->setDataLayout("");
+  }
+  if (TargetTriple.getTriple().empty())
+    TargetTriple.setTriple(DefaultTriple);
+  TargetTriple.setArch(TargetTriple.getArch(), Triple::SPIRVSubArch_v16);
+  M->setTargetTriple(TargetTriple);
+
+  int FD = -1;
+  if (std::error_code EC = sys::fs::openFileForWrite(OutputFile, FD))
+    return errorCodeToError(EC);
+  auto OS = std::make_unique<llvm::raw_fd_ostream>(FD, true);
+
+  std::string Result, ErrMsg;
+  // List of allowed extensions. Currently, all extensions are allowed.
+  // TODO: Update list of allowed extensions for SYCL compilation flow.
+  static const std::vector<std::string> AllowExtNames{"all"};
+  // Translate the Module into SPIR-V
+  if (!SPIRVTranslateModule(M.release(), Result, ErrMsg, AllowExtNames, {}))
```

Comment:
Why don't we use standard Target API to invoke CodeGen for SPIR-V target?

Reply:
After discussion with Slava, I updated it here. f446dcfbc4c747332dc7f04cacb37e8701692dca

Thanks